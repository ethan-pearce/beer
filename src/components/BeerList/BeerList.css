.beerList {
    list-style: none;
    display: grid;
    grid-template-columns: repeat(3, 33%);
    padding: 0;
}

@media only screen and (max-width: 600px) {
    .beerList {
        list-style: none;
        display: grid;
        grid-template-columns: repeat(2, 50%);
        padding: 0;
    }
  }

.beerItem {
    margin-bottom: 40px;
}

.beerItem:hover {
    opacity: 0.7;
}

.beerName{
    font-size: 24px;
    margin: 0;
}

.beerStyle{
    font-size: 16px;
    margin: 0 0 16px;
}

.beerImage{
    height: 140px;
}