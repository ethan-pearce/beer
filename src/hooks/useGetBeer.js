import { useState, useEffect } from 'react'
import { mapBeer } from '../Utils/mapBeer'

export const useGetBeer = (beerId) => {
    const [beer, setBeer] = useState([])
    const [loading, setLoading] = useState(true)
    const [error, setError] = useState(false)

    useEffect(() => {
        setLoading(true)

        fetch(`https://sandbox-api.brewerydb.com/v2/beer/${beerId}?withBreweries=Y&withSocialAccounts=Y&withIngredients=Y&key=816ecbe13a87ec829cd5aef047ca5c0a`)
            .then((res) => {
                if (!res.ok) {
                    setError(true)
                }
                return res.json()
            })
            .then((json) => {
                setBeer(mapBeerWithExtras(json.data))
                setLoading(false)
            })
            .catch(() => {
                setError(true)
                setLoading(false)
            })
    }, [beerId])

    return {
        beer,
        loading,
        error,
    }
}

const mapBeerWithExtras = (beer) => {
    return {
        ...mapBeer(beer),
        breweries: mapBreweries(beer.breweries)
    }
}

const mapBreweries = (breweries) => {
    const brews = breweries.map(brewery => {
        const { id, name, website, images, locations } = brewery;

        return {
            name,
            id,
            href: website,
            image: images.medium,
            locations: locations ? mapBreweryLocations(locations) : []
        }
    })

    return brews;
}

const mapBreweryLocations = (locations) => {
    const locat = locations.map(location => {
        const { id, name, locationTypeDisplay, region, postalCode } = location;

        return {
            name,
            id,
            type: locationTypeDisplay,
            region,
            postalCode
        }
    })

    return locat;
}