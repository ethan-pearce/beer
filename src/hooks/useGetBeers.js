import { useState, useEffect } from 'react'
import { mapBeer } from '../utils/mapBeer'

export const useGetBeers = () => {
    const [beers, setBeers] = useState([])
    const [loading, setLoading] = useState(true)
    const [error, setError] = useState(false)

    useEffect(() => {
        setLoading(true)

        fetch('https://sandbox-api.brewerydb.com/v2/beers?key=816ecbe13a87ec829cd5aef047ca5c0a')
            .then((res) => {
                if (!res.ok) {
                    setError(true)
                }
                return res.json()
            })
            .then((json) => {
                setBeers(mapBeers(json.data))
                setLoading(false)
            })
            .catch(() => {
                setError(true)
                setLoading(false)
            })
    }, [])

    return {
        beers,
        loading,
        error,
    }
}

const mapBeers = (beers) => {
    const mappedBeers = beers.map(beer => {
        return mapBeer(beer);
    })

    return mappedBeers;
}