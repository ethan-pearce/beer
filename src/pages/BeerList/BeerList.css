.container{
    margin: 32px 0;
    min-height: 600px;
}

.beerList {
    display: grid;
    grid-template-columns: repeat(3, 33%);
    padding: 16px;
}

@media only screen and (max-width: 600px) {
    .beerList {
        display: grid;
        grid-template-columns: repeat(2, 50%);
    }
  }

.beerItem {
    margin-bottom: 40px;
}

.beerItem:hover {
    opacity: 0.7;
}

.beerImage{
    height: 140px;
    float: left;
    padding-left: 4px;
}