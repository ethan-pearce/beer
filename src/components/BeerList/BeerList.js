import './BeerList.css';
import { useGetBeers } from '../../hooks/useGetBeers'
import { Link } from "react-router-dom";
import BeerHeader from '../BeerHeader/BeerHeader'

const BeerList = () => {
    const { beers, loading, error } = useGetBeers();

    return (
        <div className={'container'}>
        {loading && <div>Loading .....</div>}
        {error && <div>Sorry something went wrong</div>}
        {beers.length >= 1 &&
            <ul className={'beerList'}>
                {beers.map(beer => {
                    const { name, image, id, style, organic } = beer;
                    return (
                        <li key={id} className={'beerItem'}>
                            <Link to={`/beer-info/${id}`}>
                                <BeerHeader title={name} subTitle={style} organic={organic} />
                                <img src={image} alt={name} className={'beerImage'}/>
                            </Link>
                        </li>
                    )
                })}
            </ul>
        }
      </div>
    );
  }
  
  export default BeerList;