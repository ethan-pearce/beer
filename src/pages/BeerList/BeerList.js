import React, { useContext } from 'react';
import './BeerList.css';
import { useGetBeers } from '../../hooks/useGetBeers'
import { Link, useParams } from "react-router-dom";
import BeerHeader from '../../components/BeerHeader/BeerHeader'
import BeerContext from '../../app/BeerContext'
import Loading from '../../components/Loading/Loading'

const BeerList = () => {
    let params = useParams();
    const { beers, info, loading, error } = useGetBeers(params?.breweryId);
    const { totalResults } = info;
    const { setBeer } = useContext(BeerContext);

    return (
        <div className={'container'}>
        {loading && <Loading />}
        {error && <div>Sorry something went wrong</div>}
        {!loading &&
            <>
            {totalResults &&
                <p className={'totalResults'}>{`${totalResults} Results`}</p>
            }
            {beers.length >= 1 &&
                <ul className={'beerList'}>
                    {beers.map(beer => {
                        const { name, image, id, style, organic } = beer;
                        return (
                            <li key={id} className={'beerItem'}>
                                <Link to={`/beer-info/${id}`} onClick={() => setBeer({ name, style, organic })}>
                                    <BeerHeader title={name} subTitle={style} organic={organic} />
                                    <img src={image} alt={name} className={'beerImage'} loading="lazy"/>
                                </Link>
                            </li>
                        )
                    })}
                </ul>
            }
            </>
        }
      </div>
    );
}

export default BeerList;