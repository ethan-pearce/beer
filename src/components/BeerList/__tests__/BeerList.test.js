import React from 'react'
import { render } from '@testing-library/react';
import BeerList from '../BeerList'
import { useGetBeers } from '../../../hooks/useGetBeers'

jest.mock('../../../hooks/useGetBeers', () => ({
    useGetBeers: jest.fn(),
}))

describe('BeerList Component', () => {
    it('should render loading', () => {
        useGetBeers.mockImplementation(() => { return { beers: [], loading: true, error: false }})
        const { container } = render(<BeerList />)
        expect(container).toMatchSnapshot()
    })

    it('should render with error', () => {
        useGetBeers.mockImplementation(() => { return { beers: [], loading: false, error: true }})
        const { container } = render(<BeerList />)
        expect(container).toMatchSnapshot()
    })
})