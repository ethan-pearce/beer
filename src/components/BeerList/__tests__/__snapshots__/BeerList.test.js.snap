// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeerList Component should render loading 1`] = `
<div>
  <div
    class="container"
  >
    <div>
      Loading .....
    </div>
  </div>
</div>
`;

exports[`BeerList Component should render with error 1`] = `
<div>
  <div
    class="container"
  >
    <div>
      Sorry something went wrong
    </div>
  </div>
</div>
`;
